lib/Infer/AliveDriver.cpp

```diff
@@ -0,0 +1,400 @@
+  IR::State src_state(t.src), tgt_state(t.tgt);
+  util::sym_exec(src_state);
+  util::sym_exec(tgt_state);
+
+  util::Errors errs;
+
+  auto &ap = src_state.returnVal();
+  auto &bp = tgt_state.returnVal();
+  auto &a = ap.first;
+  auto &b = bp.first;
+
+  auto qvars = src_state.getQuantVars();
+  qvars.insert(ap.second.begin(), ap.second.end());
+
+  auto err = [&](const smt::Result &r, bool print_var, const char *msg) {
+    tools::error(errs, src_state, tgt_state, r, print_var, nullptr, a, b, msg,
+          false);
+  };
+
+  std::set<smt::expr> vars;
+  std::set<smt::expr> consts;
+
+  for (auto &[var, val] : src_state.getValues()) {
+    auto &name = var->getName();
+    if (startsWith("%var", name)) {
+      auto app = val.first.value.isApp();
+      assert(app);
+      vars.insert(Z3_get_app_arg(smt::ctx(), app, 1));
+    }
+  }
+  for (auto &[var, val] : tgt_state.getValues()) {
+    auto &name = var->getName();
+    if (startsWith("%reserved", name)) {
+      auto app = val.first.value.isApp();
+      assert(app);
+      consts.insert(Z3_get_app_arg(smt::ctx(), app, 1));
+    }
+  }
+
+  auto SimpleConstExistsCheck = smt::expr::mkForAll(vars, a.value == b.value);
+
+  smt::Solver::check({{preprocess(t, qvars, ap.second,
+                std::move(SimpleConstExistsCheck)),
+  [&] (const smt::Result &r) {
+    if (r.isUnsat()) {
+      err(r, true, "Value mismatch");
+    } else if (r.isSat()) {
+      result = r.getModel().getInt(*consts.begin());
+      return;
+    } else {
+      err(r, true, "Unknown/Invalid Result, investigate.");
+    }
+  }}});
+
+  return errs;
+}
+
+souper::AliveDriver::AliveDriver(Inst *LHS_, Inst *PreCondition_)
+    : LHS(LHS_), PreCondition(PreCondition_) {
+  InstNumbers = 101;
+  //FIXME: Magic number. 101 is chose arbitrarily.
+  //This should go away once non-input variable names are not discarded
+
+  if (!translateRoot(LHS, PreCondition, LHSF, LExprCache)) {
+    llvm::report_fatal_error("Failed to translate LHS.\n");
+  }
+}
+
+//TODO: Return an APInt when alive supports it
+std::optional<int64_t>
+souper::AliveDriver::synthesizeConstant(souper::Inst *RHS) {
+  InstNumbers = 0;
+  std::set<std::string> Consts;
+    getReservedConsts(RHS, Consts);
+  assert(!Consts.empty());
+  if (Consts.size() == 1) {
+    RExprCache.clear();
+    IR::Function RHSF;
+    if (!translateRoot(RHS, nullptr, RHSF, RExprCache)) {
+      llvm::errs() << "Failed to translate RHS.\n";
+      // TODO: Eventually turn this into an assertion
+      return std::nullopt;
+    }
+    tools::Transform t;
+    ReturnLHSRAII foo{t, LHSF};
+    t.src = std::move(LHSF);
+    t.tgt = std::move(RHSF);
+    tools::TransformVerify tv(t, /*check_each_var=*/false);
+    int64_t result = 0;
+    if (auto errs = synthesizeConstantUsingSolver(t, result)) {
+      return std::nullopt; // TODO: Encode errs into ErrorCode
+    } else {
+      return {result};
+    }
+  } else {
+    llvm::errs() << "Multiple constants unimplemented.";
```

Comment:
This can not be a fatal error as we still want those RHS guesses to be generated for when we support multiple constants.